### PR TITLE
Add switch for trim out ConfigurationManager support

### PR DESCRIPTION
**Pull request by kant2002**
With this change you can shave 900Kb out of 25.8Mb from sample SqlClient app from here https://github.com/dotnet/SqlClient/issues/1941#issuecomment-1454973332

Other example is that usage of `System.Runtime.Caching.MemoryCache` as shown in https://learn.microsoft.com/en-us/dotnet/api/system.runtime.caching.memorycache

Applying configuration switch reduce app size by 5.2Mb (from 7.8Mb to 2.6Mb)

Most likely modern applications which are first to use NativeAOT would not need this functionality because they use other means for configuration, so would be great to have this disabled.

Low drop in size in SqlClient coming from fact that there other large dependencies which keep Xml related code around.

Discovered as part of https://github.com/dotnet/SqlClient/issues/1942

### CODE REVIEW

**Comment by eerhardt** [comment]
I wonder if it would be better if we changed these "up stack" libraries to no longer use ConfigurationManager. Cutting it off inside ConfigurationManager is probably going to lead to "random" runtime errors when some untested path calls into ConfigurationManager, and it throws.

File: src/libraries/System.Configuration.ConfigurationManager/src/System.Configuration.ConfigurationManager.csproj [hunk 1]
```diff
@@ -13,8 +13,14 @@
   <!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
   <PropertyGroup>
     <IsPartialFacadeAssembly Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETFramework'">true</IsPartialFacadeAssembly>
+    <!-- ILLinker settings -->
+    <ILLinkDirectory>$(MSBuildThisFileDirectory)ILLink\</ILLinkDirectory>
```
Comment (eerhardt) [inline src/libraries/System.Configuration.ConfigurationManager/src/System.Configuration.ConfigurationManager.csproj:17]: You shouldn't need this. It is already set in 

https://github.com/dotnet/runtime/blob/340d1f22ef3b29b8c6b5369b68e73e5f7afea79c/eng/illink.targets#L10

File: src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/ConfigurationManager.cs [hunk 2]
```diff
@@ -243,6 +252,15 @@ private static void PreloadConfigurationSectionGroup(ConfigurationSectionGroup s
                 PreloadConfigurationSectionGroup(childSectionGroup);
         }
 
+        private static bool s_disableConfigurationManager;
+        private static bool DisableConfigurationManager
+        {
+            get
+            {
+                return AppContext.TryGetSwitch(@"Switch.System.Configuration.ConfigurationManager.DisableConfigurationManager", out s_disableConfigurationManager) ? s_disableConfigurationManager : false;
```
Comment (eerhardt) [inline src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/ConfigurationManager.cs:260]:
```suggestion
                return AppContext.TryGetSwitch(@"System.Configuration.ConfigurationManager.DisableConfigurationManager", out s_disableConfigurationManager) ? s_disableConfigurationManager : false;
```

We don't use the `Switch.` prefix for feature switches. See https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/feature-switches.md

**Comment by kant2002** [comment]
Oops! I definitely go overboard with some changes. My idea was to always return null when call `GetSection()` pretending that section does not exists. I start adding other changes for consistency and break things.

`GetSection()` returning null should be safe change. You already can have situation when you config is missing, and usually that mean that application use defaults. `OpenXXXConfiguration` throwing is debatable, but I put them only for consistency. Really can ignore these changes.

Regarding your question. In specific case of SqlClient the libraries which require `ConfigurationManager.GetSection()` is at least `System.Runtime.Caching` and some internal to SqlClient stuff. I think that I can do changes there, but how to keep them coordinated? That's how I end up with this change. 
I do not see that making `System.Runtime.Caching` to not depend on ConfgurationManager possible (breaking change), and do not see how MemoryCache can be replaced in SqlClient into other alternatives given that they support netfx (also breaking change).

**Review by eerhardt — COMMENTED** [review]
I think we should consider removing dependencies upstack on ConfigurationManager. It is a legacy component that we haven't been making investments in. See https://github.com/dotnet/runtime/blob/main/src/libraries/System.Configuration.ConfigurationManager/README.md

> This library exists only to support migrating existing .NET Framework code that uses System.Configuration.

**Comment by kant2002** [comment]
From here https://learn.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-5.0#systemruntimecachingmemorycache-1

> Use System.Runtime.Caching/MemoryCache as a compatibility bridge when porting code from ASP.NET 4.x to ASP.NET Core.

I would like to remove `ConfigurationManager` from that package, but seems to be that can break migration efforts to ASP.NET Core.

**Comment by roji** [comment]
@kant2002 see [my comment](https://github.com/dotnet/SqlClient/issues/1942#issuecomment-1458700518) in the SqlClient repo. tl;dr I think SqlClient needs to have a "slim" entry point for size-sensitive users, rather than introduce feature switches around ConfigurationManager (and other components).

**Comment by steveharter** [comment]
@kant2002 thanks for the PR but closing based on feedback

>Cutting it off inside ConfigurationManager is probably going to lead to "random" runtime errors when some untested path calls into ConfigurationManager, and it throws.

>  I think SqlClient needs to have a "slim" entry point for size-sensitive users, rather than introduce feature switches around ConfigurationManager (and other components).

**Pull request closed without merging**